### PR TITLE
Tweak E0308 on opaque types

### PR DESCRIPTION
**Pull request by estebank**
```
error[E0308]: if and else have incompatible types
  --> file.rs:21:9
   |
18 | /     if true {
19 | |         thing_one()
   | |         ----------- expected because of this
20 | |     } else {
21 | |         thing_two()
   | |         ^^^^^^^^^^^ expected opaque type, found a different opaque type
22 | |     }.await
   | |_____- if and else have incompatible types
   |
   = note: expected type `impl std::future::Future` (opaque type)
              found type `impl std::future::Future` (opaque type)
   = note: distinct uses of `impl Trait` result in different opaque types
   = help: if both futures resolve to the same type, consider `await`ing on both of them
```

r? @Centril
CC #63167

### CODE REVIEW

File: src/librustc/ty/error.rs [hunk 1]
```diff
@@ -247,20 +247,32 @@ impl<'tcx> ty::TyS<'tcx> {
...
                     db.note("no two closures, even if identical, have the same type");
                     db.help("consider boxing your closure and/or using it as a trait object");
                 }
+                if expected_str == found_str && expected_str == "opaque type" { // Issue #63167
+                    db.note("distinct uses of `impl Trait` result in different opaque types");
+                    let e_str = values.expected.to_string();
+                    let f_str = values.found.to_string();
+                    if &e_str == &f_str && &e_str == "impl std::future::Future" {
+                        // FIXME: use non-string based check.
+                        db.help("if both futures resolve to the same type, consider `await`ing \
```
Comment (Centril) [inline src/librustc/ty/error.rs]:
```suggestion
                        db.help("if both `Future`s have the same `Output` type, consider `.await`ing \
```

File: src/librustc/ty/error.rs [hunk 2]
```diff
@@ -247,20 +247,32 @@ impl<'tcx> ty::TyS<'tcx> {
...
                 if expected_str == found_str && expected_str == "closure" {
                     db.note("no two closures, even if identical, have the same type");
                     db.help("consider boxing your closure and/or using it as a trait object");
                 }
+                if expected_str == found_str && expected_str == "opaque type" { // Issue #63167
+                    db.note("distinct uses of `impl Trait` result in different opaque types");
+                    let e_str = values.expected.to_string();
+                    let f_str = values.found.to_string();
+                    if &e_str == &f_str && &e_str == "impl std::future::Future" {
+                        // FIXME: use non-string based check.
```
Comment (Centril) [inline src/librustc/ty/error.rs]: Is this for another PR or do you want to change it here?

**Review by Centril — COMMENTED** [review]
Code looks OK aside from the FIXME -- some tests would be nice tho :)

File: src/librustc/ty/error.rs (hunk 2 above)
Comment (estebank) [reply src/librustc/ty/error.rs]: I want to leave that as follow up work, which is why I'm not closing #63167.

File: src/librustc/ty/error.rs (hunk 1 above)
Comment (estebank) [reply src/librustc/ty/error.rs]: I'm not entirely happy with the `help`, but it's a relatively good stop gap for now. I'd like the experience in 1.39 to be as good as possible, so I'll make a round for the async/await diagnostics tickets to, if not fix them, have a stop gap like this one.

Comment (Centril) [reply src/librustc/ty/error.rs]: Sure, if you don't think my suggestion is an improvement then let's skip it... but do you? :P

**Comment by Centril** [comment]
r=me rollup with a test that exercises the behavior we're adding (e.g. from the issue) and with or without changing the message per https://github.com/rust-lang/rust/pull/63796#discussion_r316350228.

File: src/librustc/ty/error.rs (hunk 1 above)
Comment (estebank) [reply src/librustc/ty/error.rs]: The suggestion is good, I'm just not too happy with not actually checking the `Output` nor the mildly confusing "consider...". But it'll do for now.

**Comment by estebank** [comment]
@bors r=Centril rollup

cleaned up an unnecessary "missing main" error

**Comment by uberjay** [comment]
>    = help: if both futures resolve to the same type, consider `await`ing on both of them

Could rustc provide that suggestion _iff_ both futures resolve to the same type? Sounds challenging to me, but I know nothing about rustc diagnostics! 😬

**Comment by estebank** [comment]
@uberjay it could but first we need to set some extra machinery around associated types to be able to do that.

**Pull request merged**